rfd/XXXXa-scoped-join-tokens.md

```diff
@@ -0,0 +1,236 @@
+concurrent join attempts to exceed the allowed `max_uses` for a token.
+
+To reduce the throughput to the backend, each auth server would queue attempts
+to update the token and flush them as a single update on a set interval. The
+interval would need to be a relatively short as it would increase provisioning
+time by the interval itself in the worst case. This way writes could be batched
+per auth server per interval instead of per provisioning attempt.
+
+Configuring a token with a `max_uses` of `0` would effectively disable usage
+limits and fallback to the typical expiration behavior tokens have today.
+
+### `tctl` subcommands
+
+In order to provision and manage scoped tokens, we will extend `tctl` with
+`scoped` variants of the existing `tokens` sub commands. The UX of these new
+sub-commands will be nearly identical to working with unscoped provisioning
+tokens. The only concrete differences are the addition of some new flags
+specific to scoped tokens and the fact that these operations themselves are
+subject to scoped access controls. Meaning that you will not be able to create
+a token for a scope that is orthogonal or ancestor to the scope of your own
+access credentials.
+
+Below are examples for each of the sub commands that will be implemented as
+part of this RFD.
+
+```bash
+# adding a scoped token that will assign provisioned resources to the
+# /staging/west scope
+$ tctl scoped tokens add --type=node --scope=/staging/west
+```
+
+```bash
+# adding a scoped token that will assign provisioned resources to the
+# /staging/west scope, automatically assign labels, and limit provisioned
+# resources to 5
+$ tctl scoped tokens add \
+  --type node \
+  --scope /staging/west \
+  --labels env=staging,hello=world \
+  --max-uses 5
+```
+
+```bash
+# removing a scoped token by name
+$ tctl scoped tokens rm <token-name>
+```
+
+```bash
+# List all scoped tokens assigning provisioned resources to descendants of the
+# target scope /staging (e.g. tokens assigning /staging/west and /staging/east
+# will all be returned). Descendant matches are also the default behavior when
+# filtering on scope
+$ tctl scoped tokens ls --scope=/staging
+```
+
+```bash
+# List all scoped tokens assigning provisioned resources to ancestors of the
+# target scope /staging/east (e.g. tokens assigning /staging/east and /staging
+# will all be returned, but /staging/west will not be)
+$ tctl scoped tokens ls --scope=/staging/east --mode=ancestor
+```
+
+### Agent Changes
+
+As mentioned earlier in this RFD, scoped tokens are meant to be backwards
+compatible for Teleport agents. Which means the existing mechanisms for
+joining, such as `join_params` in `teleport.yaml` or the
+`--token` flag of `teleport start`, will continue to work in the same way. The
+issued host certificate will be ammended with an `AgentScope` field which will
+be used by the Teleport Auth Service during access control decisions.
```

Comment:
Are the join token names going to be namespaced? Are we going to implicitly treat regular join tokens as existing in the root scope? How do we handle name collisions?